File: src/csharp/VRE.Vridge.API.Client/Remotes/HeadRemote.cs
using System;
using System.Collections.Generic;
using System.Text;
using VRE.Vridge.API.Client.Proxy.HeadTracking;

namespace VRE.Vridge.API.Client.Remotes
{
    public class HeadRemote : RemoteBase<HeadTrackingProxy>
    {
        internal HeadRemote(HeadTrackingProxy proxy) : base(proxy)
        {
            
        }

        /// <summary>
        /// Reorients tracking system and sets new center to current head direction.
        /// </summary>
        public void Recenter()
        {
            WrapTimeouts(() => Proxy.RecenterView());
        }

        public void SetPosition(float x, float y, float z)
        {
            WrapTimeouts(() => Proxy.SetPosition(x, y, z));
        }

        public void SetRotationAndPosition(float yaw, float pitch, float roll, float x, float y, float z)
        {
            WrapTimeouts(() => Proxy.SetRotationAndPosition(yaw, pitch, roll, x, y, z));
        }
        
        public void SetAsyncOffset(float yaw)
        {
            WrapTimeouts(() => Proxy.SetAsyncOffset(yaw));
        }

        public float[] GetCurrentPose()
        {
            return WrapTimeouts(() => Proxy.GetCurrentPhonePose());
        }

        public void SetStatus(bool isInTrackingRange)
        {
            WrapTimeouts(() => Proxy.ChangeTrackingState(isInTrackingRange));
        }

        // Type-cast-methods
        public void SetAsyncOffset(double yaw) => 
            SetAsyncOffset((float)yaw);

        public void SetRotationAndPosition(double yaw, double pitch, double roll, double x, double y, double z) =>
            SetRotationAndPosition((float) yaw, (float) pitch, (float) roll, (float) x, (float) y, (float) z);

        public void SetPosition(double x, double y, double z) =>
            SetPosition((float) x, (float) y, (float) z);

        internal override void Dispose()
        {
            Proxy?.Disconnect();
            base.Dispose();
        }        
    }
}

File: src/java/VRE.Vridge.API/VRE.Vridge.API.Client/src/main/java/com/riftcat/vridge/api/client/java/remotes/HeadRemote.java
package com.riftcat.vridge.api.client.java.remotes;

import com.riftcat.vridge.api.client.java.proto.TrackedPose;
import com.riftcat.vridge.api.client.java.proxy.HeadTrackingProxy;

import java.util.concurrent.TimeoutException;

public class HeadRemote extends RemoteBase {
    private HeadTrackingProxy proxy;

    HeadRemote(HeadTrackingProxy proxy) {
        super(proxy);

        this.proxy = proxy;
    }

    /**
     * Reorients tracking system and sets new center to current head direction.
     */
    public void recenter(){
        try{
            proxy.recenterView();
        }
        catch (Exception x){
            dispose();
        }

    }

    /**
     * Gets current head pose and related offsets. May return null on connection issues.
     */
    public TrackedPose getCurrentPose() {

        try{
            return proxy.getCurrentPose();
        }
        catch (Exception e){
            dispose();
            return null;
        }
    }

    /**
     * Sets head position to new location.
     */
    public void setPosition(float x, float y, float z){
        try{
            proxy.setPosition(x, y, z);
        }
        catch (Exception e){
            dispose();
        }
    }

    /**
     * Sets head position to new location and orientation.
     * This won't work for headsets with reprojection enabled.
     */
    public void setRotationAndPosition(float yaw, float pitch, float roll, float x, float y, float z){
        try{
            proxy.setRotationAndPosition(yaw, pitch, roll, x, y, z);
        }
        catch (Exception e){
            dispose();
        }
    }


    /**
     * Sets offsets in yaw axis to be applied to each head and controller pose processed by the system.
     * @param yaw Offset in radians.
     */
    public void setYawOffset(float yaw){
        try{
            proxy.setYawOffset(yaw);
        }
        catch (Exception x){
            dispose();
        }
    }

    /**
     * Marks the headset as in/outside of tracking range. Setting it to false will most likely
     * stop rendering on SteamVR side as pose data will be considered invalid.
     */
    public void setStatus(boolean isInTrackingRange)
    {
        try{
            proxy.changeTrackingState(isInTrackingRange);
        }
        catch (Exception e){
            dispose();
        }
    }

    // Type-cast-methods
    /**
     * Sets head position to new location and orientation.
     * This won't work for headsets with reprojection enabled.
     */
    public void setRotationAndPosition(double yaw, double pitch, double roll, double x, double y, double z){
        setRotationAndPosition((float) yaw, (float) pitch, (float) roll, (float) x, (float) y, (float) z);
    }

    /**
     * Sets head position to new location.
     */
    public void setPosition(double x, double y, double z) {
        setPosition((float) x, (float) y, (float) z);
    }
}
